plugins/kotlin/uast/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/internal/firKotlinInternalUastUtils.kt

```diff
@@ -56,7 +57,9 @@ internal fun KtAnalysisSession.toPsiType(ktType: KtType, context: KtElement): Ps
             StandardClassIds.Char -> PsiType.CHAR
             StandardClassIds.Double -> PsiType.DOUBLE
             StandardClassIds.Float -> PsiType.FLOAT
-            StandardClassIds.Unit -> PsiType.VOID
+            StandardClassIds.Unit -> {
+                if (context.isFunctionWithImplicitUnitReturnType) PsiType.VOID else null
```

Comment:
If we are using context like this, are we sure that context is always a target declaration? By the code it looks like that context is some random `PsiElement` which we use to get the module in which it lives

Reply:
This `context` is where the given `KtType` for FIR UAST (or `KotlinType` for FE 1.0 UAST) is retrieved. Therefore, if this `context` is a function with implicit `Unit` return type, it came from either
* https://github.com/JetBrains/intellij-community/blob/kt-212-master/plugins/kotlin/uast/uast-kotlin/src/org/jetbrains/uast/kotlin/KotlinUastResolveProviderService.kt#L261 or
* https://github.com/JetBrains/intellij-community/blob/kt-212-master/plugins/kotlin/uast/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/FirKotlinUastResolveProviderService.kt#L278

where the `context` is the declaration of interest.

I added a test to verify behaviors; see my other comments.

Reply:
Let's not relay that the `context` is always the declaration type came from. Some new usages of `toPsiType` may be added which breaks this rule. If you would like to use `context` like this, I would suggest renaming and changing it type to `declaration: KtDeclaration` so we ensure that we always pass owner declaration. Or maybe provide some additional enum/flag specifying type conversion mode for `Unit`.

Reply:
These type conversions are used at various places, so it's not trivial to change the type of `context` to `KtDeclaration`. Instead, I prefer the idea of an additional enum (named `TypeOwnerKind`) to specify the kind of type owner, which may be different from the `context` of this type conversion. Btw, sometimes this `context` is used to retrieve the project, module, etc. as you mentioned.

In FIR UAST, type owner is always that `context`, so quite straightforward. In FE 1.0 UAST, it's quite similar for many cases, but for descriptor-based fake declaration or declaration from deserialization, it's unclear whether the `context` is really the owner of the type to be converted. Hence, hard-coded `DECLARATION` type owner kind for those cases.

Reply:
I meant something like 

```
KtAnalysisSession.toPsiType(
    ktType: KtType,
    context: KtElement,
    typeConversionMode: TypeConversionMode,
    boxed: Boolean = false
): PsiType {
  ...
   if (typeConversionMode.convertUnitToVoid) PsiType.VOID.orBoxed() else null
   ...
}

enum class TypeConversionMode(val convertUnitToVoid: Boolean) {
	...
}
```

But your solution works well too :)